File: LeetCode/Java/Easy/SubtreeOfAnotherTree.java
package Easy;

public class SubtreeOfAnotherTree {
	
	public static void main(String[] args){
		
	}
	
	public static boolean isSubtree(TreeNode s, TreeNode t){	
		if (s == null && t == null){
			return true;
		}
		if (s == null || t== null){
			return false;
		}
		if (s.val == t.val){
			
			return isPerfectSubtree(s,t) || isSubtree(s.left, t) || isSubtree(s.right, t);
		}
		
		return isSubtree(s.left, t) || isSubtree(s.right, t);
	}
	
	public static boolean isPerfectSubtree(TreeNode s, TreeNode t){
		if (s == null && t == null){
			return true;
		}
		if (s == null || t== null){
			return false;
		}
		if (s.val != t.val){
			return false;
		}
		
		return isPerfectSubtree(s.right, t.right) && isPerfectSubtree(s.left, t.left);
	}

}

File: LeetCode/Python/Easy/SubtreeOfAnotherTree.py
class Solution(object):
    def isSubtree(self, s, t):
        if s is None:
            return False

        if s.val == t.val:
            if self.checkTree(s, t):
                return True

        return self.isSubtree(s.right, t) or self.isSubtree(s.left, t)

    def checkTree(self, s, t):
        if s is None and t is None:
            return True
        if s is None or t is None:
            return False
        if s.val == t.val:
            return self.checkTree(s.left, t.left) and self.checkTree(s.right, t.right)
